src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpAuthHelper.cs

```diff
@@ -0,0 +1,299 @@
+                    if (!Interop.WinHttp.WinHttpQueryAuthSchemes(
+                        state.RequestHandle,
+                        out supportedSchemes,
+                        out firstSchemeIgnored,
+                        out authTarget))
+                    {
+                        WinHttpException.ThrowExceptionUsingLastError();
+                    }
+
+                    // WinHTTP returns the proper authTarget based on the status code (401, 407).
+                    // But we can validate with assert.
+                    Debug.Assert(authTarget == Interop.WinHttp.WINHTTP_AUTH_TARGET_PROXY);
+
+                    proxyAuthScheme = ChooseAuthScheme(supportedSchemes);
+                    state.RetryRequest = true;
+                    break;
+
+                default:
+                    if (state.PreAuthenticate && serverAuthScheme != 0)
+                    {
+                        SaveServerCredentialsToCache(uri, serverAuthScheme, state.ServerCredentials);
+                    }
+                    break;
+            }
+        }
+
+        public void PreAuthenticateRequest(WinHttpRequestState state, uint proxyAuthScheme)
+        {
+            // Set proxy credentials if we have them.
+            // If a proxy authentication challenge was responded to, reset
+            // those credentials before each SendRequest, because the proxy  
+            // may require re-authentication after responding to a 401 or  
+            // to a redirect. If you don't, you can get into a 
+            // 407-401-407-401- loop.
+            if (proxyAuthScheme != 0)
+            {
+                SetWinHttpCredential(
+                    state.RequestHandle,
+                    state.Proxy == null ? state.DefaultProxyCredentials : state.Proxy.Credentials,
+                    state.RequestMessage.RequestUri,
+                    proxyAuthScheme,
+                    Interop.WinHttp.WINHTTP_AUTH_TARGET_PROXY);
+            }
+
+            // Apply pre-authentication headers for server authentication?
+            if (state.PreAuthenticate)
+            {
+                uint authScheme;
+                NetworkCredential serverCredentials;
+                if (GetServerCredentialsFromCache(
+                    state.RequestMessage.RequestUri,
+                    out authScheme,
+                    out serverCredentials))
+                {
+                    SetWinHttpCredential(
+                        state.RequestHandle,
+                        serverCredentials,
+                        state.RequestMessage.RequestUri,
+                        authScheme,
+                        Interop.WinHttp.WINHTTP_AUTH_TARGET_SERVER);
+                    state.LastStatusCode = HttpStatusCode.Unauthorized; // Remember we already set the creds.
+                }
+            }
+        }
+
+        public bool GetServerCredentialsFromCache(
+            Uri uri,
+            out uint serverAuthScheme,
```

Comment:
Design: The presence of multiple `out` parameters is a code smell. The code is having too many responsibilities: we would have more clarity if we extracted the Cache in a different class.